plugin/plugin_test.go

```diff
@@ -22,115 +23,110 @@ func TestConfigurationParser(t *testing.T) {
-		}
-		wg.Add(1)
-		go func(wg *sync.WaitGroup, mut *sync.Mutex, i int) {
-			defer wg.Done()
-			for j := 0; j < 1; j++ {
-				resp, err := svr.Client().Get(requestURL)
-				if err != nil {
-					fmt.Printf("Error: %s", err)
-					// t.Errorf("Error in %v, expected: %v, got: %v", test.url, test.expectedStatusCode, err.Error())
-				}
-				if resp.StatusCode == 200 {
-					mut.Lock()
-					results = append(results, fmt.Sprintf("PASS: i=%v, j=%v, time=%v", i, j, time.Since(initialTime).Milliseconds()))
-					mut.Unlock()
-				} else {
-					mut.Lock()
-					results = append(results, fmt.Sprintf("FAIL: i=%v, j=%v, time=%v", i, j, time.Since(initialTime).Milliseconds()))
-					mut.Unlock()
+	ticker := time.NewTicker(time.Second * 1)
+
+	i := 0
+
+	for {
+		<-ticker.C
+
+		for i := 0; i < 200; i++ {
+
+			wg.Add(1)
+			go func(wg *sync.WaitGroup, i int) {
+				defer wg.Done()
+				for j := 0; j < 1; j++ {
+					resp, err := svr.Client().Get(requestURL)
+					if err != nil {
+						t.Errorf("Error in %v, expected: %v, got: %v", i, 200, err.Error())
+					}
+					assert.Equal(t, 200, resp.StatusCode)
 				}
-			}
-		}(wg, mut, i)
+			}(wg, i)
+		}
+
+		i++
+
+		if i == 5 {
+			ticker.Stop()
+			break
+		}
+
 	}
+
 	wg.Wait()
-	prettyPrint(results)
 }
 
-// 1000 requests in a second
+// 1000 requests in a second should be executed successfully.
+// currently it is taking 625ms to execute 1000req/sec
 func TestStressOfRateLimit(t *testing.T) {
 	wg := &sync.WaitGroup{}
 
-	results := []string{}
-	mut := &sync.Mutex{}
-	initialTime := time.Now()
-
 	// get an instance of http test server with waf
-	conf := `SecRule ARGS:id "@eq 1" "id:1, ratelimit:zone[]=%{REQUEST_HEADERS.host}&events=300&window=1&interval=2&action=deny&status=401, pass, status:200"`
+	conf := `SecRule ARGS:id "@eq 1" "id:1, ratelimit:zone[]=%{REQUEST_HEADERS.host}&events=300&window=1&interval=2&action=deny&status=429, pass, status:200"`
 
 	svr := helpers.NewHttpTestWafServer(conf)
 	defer svr.Close()
 
 	requestURL := fmt.Sprintf("%v?id=1", svr.URL)
 
+	currentTime := time.Now().UnixMilli()
+
 	for i := 0; i < 1000; i++ {
+
 		wg.Add(1)
-		go func(wg *sync.WaitGroup, mut *sync.Mutex, i int) {
+		go func(wg *sync.WaitGroup, i int) {
 			defer wg.Done()
 			for j := 0; j < 1; j++ {
-				resp, err := svr.Client().Get(requestURL)
+				_, err := svr.Client().Get(requestURL)
 				if err != nil {
-					fmt.Printf("Error: %s", err)
-					// t.Errorf("Error in %v, expected: %v, got: %v", test.url, test.expectedStatusCode, err.Error())
-				}
-				if resp.StatusCode == 200 {
-					mut.Lock()
-					results = append(results, fmt.Sprintf("PASS: i=%v, j=%v, time=%v", i, j, time.Since(initialTime).Milliseconds()))
-					mut.Unlock()
-				} else {
-					mut.Lock()
-					results = append(results, fmt.Sprintf("FAIL: i=%v, j=%v, time=%v", i, j, time.Since(initialTime).Milliseconds()))
-					mut.Unlock()
+					t.Errorf("Error in %v, expected: %v, got: %v", i, 200, err.Error())
```

Comment:
Not sure this will work, I would first do `t.Fatal` and also use a channel for catching errors at the end.
Reference https://github.com/ipfs/kubo/issues/2043